File: backend/src/lib/email.ts
import nodemailer from "nodemailer";
import { Resend } from "resend";

let transporter: nodemailer.Transporter | null = null;

const SMTP_CONNECTION_TIMEOUT_MS = 15_000;
const SMTP_GREETING_TIMEOUT_MS = 10_000;

function getSmtpConfig() {
  return {
    host: process.env.SMTP_HOST || "smtp.hostinger.com",
    port: parseInt(process.env.SMTP_PORT || "465", 10),
    secure: process.env.SMTP_SECURE !== "false",
    user: process.env.SMTP_USER,
    pass: process.env.SMTP_PASS,
    from: process.env.EMAIL_FROM || process.env.SMTP_USER,
  };
}

function getTransporter(): nodemailer.Transporter | null {
  const config = getSmtpConfig();
  if (transporter) return transporter;

  if (!config.user || !config.pass) {
    console.warn("[Email] SMTP no configurado: faltan SMTP_USER o SMTP_PASS en .env. Los correos no se enviarán.");
    return null;
  }

  transporter = nodemailer.createTransport({
    host: config.host,
    port: config.port,
    secure: config.secure,
    auth: { user: config.user, pass: config.pass },
    connectionTimeout: SMTP_CONNECTION_TIMEOUT_MS,
    greetingTimeout: SMTP_GREETING_TIMEOUT_MS,
  });

  return transporter;
}

/** Si existe RESEND_API_KEY se usa Resend (HTTPS); en Render evita el bloqueo de SMTP. */
function useResend(): boolean {
  return Boolean(process.env.RESEND_API_KEY?.trim());
}

export interface SendEmailResult {
  success: boolean;
  messageId?: string;
  error?: string;
}

function getFromAddress(): string {
  return process.env.EMAIL_FROM || process.env.SMTP_USER || "onboarding@resend.dev";
}

async function sendEmailResend(
  to: string | string[],
  subject: string,
  htmlBody: string
): Promise<SendEmailResult> {
  const apiKey = process.env.RESEND_API_KEY?.trim();
  if (!apiKey) {
    return { success: false, error: "RESEND_API_KEY no configurado" };
  }

  const from = getFromAddress();
  const toList = Array.isArray(to) ? to : [to];
  const resend = new Resend(apiKey);

  try {
    const { data, error } = await resend.emails.send({
      from: `"La voz de las páginas" <${from}>`,
      to: toList,
      subject,
      html: htmlBody,
    });

    if (error) {
      console.error("[Email] Resend error:", { to: toList, subject, error });
      return { success: false, error: error.message };
    }
    console.log("[Email] Enviado (Resend):", { to: toList, subject, messageId: data?.id });
    return { success: true, messageId: data?.id };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    console.error("[Email] Resend exception:", { to: toList, subject, error: message });
    return { success: false, error: message };
  }
}

/**
 * Envía un correo.
 * - Con RESEND_API_KEY: usa Resend (recomendado en Render; no depende de SMTP).
 * - Sin él: usa SMTP (Hostinger, etc.). En Render los puertos SMTP suelen estar bloqueados.
 */
export async function sendEmail(
  to: string | string[],
  subject: string,
  htmlBody: string
): Promise<SendEmailResult> {
  if (useResend()) {
    return sendEmailResend(to, subject, htmlBody);
  }

  const trans = getTransporter();
  if (!trans) {
    console.warn("[Email] Envío omitido (SMTP no configurado):", { to, subject });
    return { success: false, error: "SMTP no configurado" };
  }

  const from = getSmtpConfig().from;
  if (!from) {
    console.error("[Email] EMAIL_FROM o SMTP_USER no definido.");
    return { success: false, error: "Remitente no configurado" };
  }

  try {
    const result = await trans.sendMail({
      from: `"La voz de las páginas" <${from}>`,
      to: Array.isArray(to) ? to.join(", ") : to,
      subject,
      html: htmlBody,
    });
    console.log("[Email] Enviado (SMTP):", { to: Array.isArray(to) ? to : [to], subject, messageId: result.messageId });
    return { success: true, messageId: result.messageId };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    console.error("[Email] SMTP error:", { to, subject, error: message });
    return { success: false, error: message };
  }
}

File: backend/app/email.py
"""Email sending: Resend (HTTPS API) when RESEND_API_KEY is set, otherwise SMTP.

Mirrors backend-old/src/lib/email.ts.
"""

import logging
from dataclasses import dataclass

import aiosmtplib
import resend
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from starlette.concurrency import run_in_threadpool

from app.config import settings

logger = logging.getLogger("app.email")

SMTP_CONNECTION_TIMEOUT_SECONDS = 15
FROM_DISPLAY_NAME = "La voz de las páginas"


@dataclass
class SendEmailResult:
    success: bool
    message_id: str | None = None
    error: str | None = None


def _use_resend() -> bool:
    return bool(settings.resend_api_key and settings.resend_api_key.strip())


def _from_address() -> str:
    return settings.email_from or settings.smtp_user or "onboarding@resend.dev"


async def _send_email_resend(to: str, subject: str, html_body: str) -> SendEmailResult:
    api_key = (settings.resend_api_key or "").strip()
    if not api_key:
        return SendEmailResult(success=False, error="RESEND_API_KEY no configurado")

    from_address = _from_address()
    resend.api_key = api_key

    try:
        response = await run_in_threadpool(
            resend.Emails.send,
            {
                "from": f'"{FROM_DISPLAY_NAME}" <{from_address}>',
                "to": [to],
                "subject": subject,
                "html": html_body,
            },
        )
        message_id = response.get("id") if isinstance(response, dict) else None
        logger.info("[Email] Enviado (Resend): to=%s subject=%s messageId=%s", to, subject, message_id)
        return SendEmailResult(success=True, message_id=message_id)
    except Exception as err:  # noqa: BLE001 - mirrors the JS try/catch around the Resend SDK call
        logger.error("[Email] Resend exception: to=%s subject=%s error=%s", to, subject, err)
        return SendEmailResult(success=False, error=str(err))


async def _send_email_smtp(to: str, subject: str, html_body: str) -> SendEmailResult:
    if not settings.smtp_user or not settings.smtp_pass:
        logger.warning("[Email] SMTP no configurado: faltan SMTP_USER o SMTP_PASS. Los correos no se enviarán.")
        return SendEmailResult(success=False, error="SMTP no configurado")

    from_address = settings.email_from or settings.smtp_user
    if not from_address:
        logger.error("[Email] EMAIL_FROM o SMTP_USER no definido.")
        return SendEmailResult(success=False, error="Remitente no configurado")

    message = MIMEMultipart("alternative")
    message["From"] = f'"{FROM_DISPLAY_NAME}" <{from_address}>'
    message["To"] = to
    message["Subject"] = subject
    message.attach(MIMEText(html_body, "html", "utf-8"))

    try:
        await aiosmtplib.send(
            message,
            hostname=settings.smtp_host,
            port=settings.smtp_port,
            use_tls=settings.smtp_secure,
            username=settings.smtp_user,
            password=settings.smtp_pass,
            timeout=SMTP_CONNECTION_TIMEOUT_SECONDS,
        )
        logger.info("[Email] Enviado (SMTP): to=%s subject=%s", to, subject)
        return SendEmailResult(success=True)
    except Exception as err:  # noqa: BLE001 - mirrors the JS try/catch around nodemailer's sendMail
        logger.error("[Email] SMTP error: to=%s subject=%s error=%s", to, subject, err)
        return SendEmailResult(success=False, error=str(err))


async def send_email(to: str, subject: str, html_body: str) -> SendEmailResult:
    if _use_resend():
        return await _send_email_resend(to, subject, html_body)
    return await _send_email_smtp(to, subject, html_body)


def verification_email_html(username: str, code: str, *, is_resend: bool = False) -> str:
    intro = "Tu nuevo código de verificación es" if is_resend else "Tu código de verificación para <strong>La voz de las páginas</strong> es"
    footer = "Si no fuiste tú, puedes ignorar este mensaje." if is_resend else "Si no creaste esta cuenta, puedes ignorar este mensaje."
    return f"""
      <h2>Hola, {username}</h2>
      <p>{intro}:</p>
      <p style="font-size:1.5rem;font-weight:700;letter-spacing:0.3em;margin:1rem 0;">{code}</p>
      <p>Introduce este código en la web{"" if is_resend else " para activar tu cuenta"}. El código caduca en 24 horas.</p>
      <p>{footer}</p>
      <p>— El equipo</p>
    """
